src/heaps/binomial_heap.jl

```diff
@@ -0,0 +1,371 @@
+# Binomial Heap
+mutable struct MutableBinomialHeapNode{T} 
+    data::T
+    degree::Int
+    handle::Int
+    parent::Union{MutableBinomialHeapNode,Nothing}
+    child::Union{MutableBinomialHeapNode,Nothing}
+    sibling::Union{MutableBinomialHeapNode,Nothing}
+end
+MutableBinomialHeapNode(data::T, handle::Int) where T = MutableBinomialHeapNode{T}(data, 0, handle, nothing, nothing, nothing)
+#################################################
+#
+#   heap type and constructors
+#
+#################################################
+mutable struct MutableBinomialHeap{T} <: AbstractHeap{T}
+    rootList::Vector{Union{Nothing,MutableBinomialHeapNode{T}}}
+    nodecount::Int
+    cumm_nodecount::Int
+    MutableBinomialHeap{T}() where T = new{T}([], 0, 0)
+
+    function MutableBinomialHeap{T}(xs::AbstractVector{T}) where T
+        BinomialHeap = _make_binomial_heap(xs, T)
+        new{T}(BinomialHeap.rootList, BinomialHeap.nodecount, BinomialHeap.cumm_nodecount)
+    end
+    
+end
+const BinomialHeap{T} = MutableBinomialHeap{T}
+BinomialHeap(xs::AbstractVector{T}) where T = MutableBinomialHeap{T}(xs)
+
+struct HeapBoundsError <: Exception
+    msg::AbstractString
+end
```

Comment:
Why now use the existing `BoundsError` ?

Reply:
Since the above implementation uses vector only for root nodes, so for the nodes other than root nodes, I had to resort to this method to show out of Bounds access.
Furthermore, I also wish to wrap the bounds error for all types of heaps, as I have opened the issue #598.